component/src/stake/component.rs

```diff
@@ -46,7 +46,11 @@ pub struct Staking {
     delegation_changes: DelegationChanges,
     /// List of changes to the tendermint validator set accumulated throughout
     /// this block, to be returned during `EndBlock`.
-    tm_validator_updates: BTreeMap<IdentityKey, u64>,
+    tm_validator_updates: BTreeMap<PublicKey, u64>,
+    /// List of updated consensus keys tracked during this block. Used to filter
+    /// the `tm_validator_updates` to avoid reporting 0 voting power for new
+    /// consensus keys.
+    new_consensus_keys: Vec<PublicKey>,
```

Comment:
I think a `BTreeSet` would be more efficient here, because we repeatedly query for inclusion. That would reduce time complexity for the query loop from quadratic to log-linear.